### PR TITLE
i61: Allow leaving group conversations 

### LINKED ISSUE #61: Allow leaving group conversations

### PR DESCRIPTION
**Pull request by nickmeinhold**
fixes #61 

Swipe to dismiss logic is from: [Implement swipe to dismiss - Flutter](https://flutter.dev/docs/cookbook/gestures/dismissible) 

Removing yourself involves: 
- adding an entry to `conversations/{cid}/leave/{userId}` 
- a cloud function triggers and updates the conversation doc 

Known Issues (all will be fixed by keeping a global state) 
- exception thrown when dismissible is not removed from the widget tree (non-fatal) 
- swiping to remove, then canceling, leaves an empty space 

I made some notes on our state management needs: [State Management · Adventures-In/chat_app Wiki](https://github.com/Adventures-In/chat_app/wiki/State-Management)

### CODE REVIEW

File: lib/models/conversations_view_model.dart [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+  void remove({@required ConversationItem item}) {
+    _items.remove(item);
+    // Tell the widgets that are listening to this model to rebuild.
+    notifyListeners();
+  }
+
+  /// Sets the list of conversations
+  void populateWith(List<ConversationItem> models) {
+    _items.clear();
+    _items.addAll(models);
```
Comment (gaslitbytech) [inline lib/models/conversations_view_model.dart:28]: This implementation makes more sense.

File: test/unit/conversations_page_test.dart [hunk 2]
```diff
@@ -31,7 +31,7 @@ void main() {
       expect(inject_list, alternate_list);
     });
 
-    test('.populateWith() does not replace _items when pre initialised', () {
+    test('.populateWith() replaces _items when already has items', () {
```
Comment (gaslitbytech) [inline test/unit/conversations_page_test.dart]: It great that writing tests is helping us remove quirks from the system with a more robust approach. Previously I had to assume there was a purpose to why it was implemented in such a way.

**Review by gaslitbytech — DISMISSED** [review]
Great work @nickmeinhold.

File: test/unit/conversations_page_test.dart (hunk 2 above)
Comment (nickmeinhold) [reply test/unit/conversations_page_test.dart]: I'm loving how much testing we're doing! We're approaching rockstar levels ;-)

**Comment by nickmeinhold** [comment]
Thanks for the approve @tourdownunder! How would you like to proceed with the merging of this and your current PR? I imagine there will be some sorting out of conflicts. I'm happy for yours to go first and handle any conflicts in this one, whatever you'd prefer is fine for me. Thanks!

**Comment by gaslitbytech** [comment]
> How would you like to proceed with the merging of this and your current PR? I imagine there will be some sorting out of conflicts. I'm happy for yours to go first and handle any conflicts in this one.

Your too kind @nickmeinhold. I'll take advantage of your generosity.

**Comment by nickmeinhold** [comment]
@tourdownunder thanks for approving this a couple of days ago. I merged the latest dev and resolved the conflicts which has auto dismissed the stale review.  If you're happy with it it's good to go... thanks!

**Comment by nickmeinhold** [comment]
Just thinking about this PR: there's a couple of things that should be tested that would be well suited to integration tests: 
- the conversation view model has been removed from the conversations page in preparation for moving the vm into the db service 
- leaving conversations involves a db change which triggers a cloud function which then makes another db change 

I've added #186 but happy to be asked to include the testing in the PR. 

Thanks!

**Review by gaslitbytech — APPROVED** [review]
Well done @nickmeinhold

**Pull request merged**